provider/pkg/schema/gen_tags.go

```diff
@@ -0,0 +1,73 @@
+package schema
+
+import (
+	"strings"
+
+	pschema "github.com/pulumi/pulumi/pkg/v3/codegen/schema"
+)
+
+type TagsStyle string
+
+const (
+	// TagsStyleUnknown indicates we can't identify the style of tags.
+	TagsStyleUnknown TagsStyle = ""
+	// TagsStyleUntyped is a style where the tags are represented as "Any" - without a schema.
+	TagsStyleUntyped TagsStyle = "untyped"
+	// TagsStyleStringMap is a style where the tags are represented as a map of strings.
+	TagsStyleStringMap TagsStyle = "stringMap"
+	// TagsStyleKeyValueArray is a style where the tags are represented as an array of key-value pairs.
+	TagsStyleKeyValueArray TagsStyle = "keyValueArray"
+)
+
+// GetTagsStyle identifies the style of tags based on the shape of the schema.
+func (ctx *context) GetTagsStyle(typeSpec *pschema.TypeSpec) TagsStyle {
+	if typeSpec == nil {
+		return TagsStyleUnknown
+	}
+	// Check for "Any" ref
+	if typeSpec.Ref == "pulumi.json#/Any" {
+		return TagsStyleUntyped
+	}
+
+	if typeSpec.AdditionalProperties != nil && typeSpec.AdditionalProperties.Type == "string" {
+		return TagsStyleStringMap
+	}
+
+	if ctx.tagStyleIsKeyValueArray(typeSpec) {
+		return TagsStyleKeyValueArray
+	}
+
+	return TagsStyleUnknown
+}
+
+func (ctx *context) tagStyleIsKeyValueArray(typeSpec *pschema.TypeSpec) bool {
+	if typeSpec == nil || typeSpec.Items == nil || typeSpec.Items.Ref == "" {
+		return false
+	}
+
+	typeToken, hasPrefix := strings.CutPrefix(typeSpec.Items.Ref, "#/types/")
+	if !hasPrefix {
+		return false
+	}
+
+	// We can't include tags which are create-only properties because this has to be added on the tags type but this is a shared type
```

Comment:
Frustrating. The way we represent create-only properties should be a property of the resource, not a type annotation!

Reply:
Agreed. It's even more annoying because when the SDKs get rendered they _are_ then just a set of properties on the resource.

Reply:
Here's an existing issue to add a better option for this in the schema: 
- https://github.com/pulumi/pulumi/issues/9158

Our other option would be to ditch using the engine's diff and instead implement this in the provider so we can customise when we do replacements. But that's quite a large piece to take on.

Reply:
What if we had `aws-native:index:Tag` and `aws-native:index:CreateOnlyTag`?

Reply:
Great minds ... that's exactly what I ended up adding for now!

[Implement shared type for all tags that cause replacement](https://github.com/pulumi/pulumi-aws-native/pull/1348/commits/e847c2c1fc6f508684441a1b591c25c8f611dcf0)